### ISSUE TITLE
Few suggestions

### ISSUE THREAD

**Issue by AngelicosPhosphoros** [auxoncorp/fixed-slice-vec#19]
I just encountered use-case where your crate can be suitable but noticed few things that should be improved.

# Correctness suggestions
## Drop probably shouldn't be implemented if elements is not drop
Code here
https://github.com/auxoncorp/fixed-slice-vec/blob/df78d1635a8694e9aad95d275daf53ed394e4a49/src/vec.rs#L22..L26

I suggest to add additional bound `T: Drop` here because it would be noop anyway after optimization.
With new bound, rustc wouldn't generate drop calls when FixedSliceVec with Copy types goes out of scope so it would have less compilation errors.

Also, it would make programs like that compile (currently you need to manually drop v before second loop):
```rust
let mut arr: [MaybeUninit<u32>; 50] = MaybeUninit::uninit_array::<50>();
let mut v: FixedSliceVec<u32> = (&mut arr[..]).into();
for i in 0..10{
   v.push(i);
}

for v in arr[..10].iter(){}
```

## `FixedSliceVec::clear` and `FixedSliceVec::truncate` can double drop
Code here
https://github.com/auxoncorp/fixed-slice-vec/blob/df78d1635a8694e9aad95d275daf53ed394e4a49/src/vec.rs#L291

If vec contains few items inside and item in the middle panics during dropping, you can call `clear` again and make double-free for first items in vec.
I suggest to move length zeroing before unsafe block so you would just leak data in case of panic.

Another option is to call `pop` until our len is 0 but it can be less performant and would make API not clear (e.g. we can still have items after panic).

Same can happen in truncate function.

## `FixedSliceVec::pop`  can be improved

Code here
https://github.com/auxoncorp/fixed-slice-vec/blob/df78d1635a8694e9aad95d275daf53ed394e4a49/src/vec.rs#L281..L282

It would be easier for optimizer to understand code like that:
```rust
self.len -= 1;
/*return*/ Some(unsafe { self.storage[self.len].as_mut_ptr().read() } )
```
It uses [`MaybeUninit`](https://doc.rust-lang.org/stable/std/mem/union.MaybeUninit.html#method.as_mut_ptr) utilities instead of slice trick.

# My own use-case suggestion

It would be useful to be able convert vec to initialized part of inner slice if elements is Copy. Since elements is Copy they cannot implement Drop so we don't need to worry about leaking elements.

Use case:
```rust
/// Returns all items which was processed.
fn do_some_processing_for_items<'a>(items: &[u32], buffer_for_output: FixedSliceVec<'a>)->&'a [u32]{
     for item in items.iter(){
         if can_process(item){
            process_item(item);
            buffer_for_output.push(*item);
         }
     }
     buffer_for_output.into_initialized_slice()
}
```

**Comment by ZackPierce** [comment]
Thank you for the suggestions, I'll look into these.

**Comment by ZackPierce** [comment]
Several of these are implemented in https://github.com/auxoncorp/fixed-slice-vec/pull/20 .

IIRC, `Drop` cannot be conditionally implemented in Rust at present, so only implementing Drop when  `T` is not  `Copy` is not in the cards for now.
